esp-hal/src/sdm.rs

```diff
@@ -0,0 +1,296 @@
+    _ref: &'d ChannelRef<'d, N>,
+    output: PeripheralRef<'d, O>,
+}
+
+impl<'d, const N: u8, O: OutputPin> Drop for Channel<'d, N, O> {
+    fn drop(&mut self) {
+        let signal = CHANNELS[N as usize];
+        self.output
+            .disconnect_from_peripheral_output(signal, private::Internal);
+    }
+}
+
+impl<'d, const N: u8, O: OutputPin> Channel<'d, N, O> {
+    /// Set raw pulse density
+    ///
+    /// Sigma-delta quantized density of one channel, the value ranges from -128
+    /// to 127, recommended range is -90 ~ 90. The waveform is more like a
+    /// random one in this range.
+    pub fn set_pulse_density(&self, density: i8) {
+        GPIO_SD::set_pulse_density(N, density);
+    }
+
+    /// Set duty cycle
+    pub fn set_duty(&self, duty: u8) {
+        let density = duty as i16 - 128;
+        self.set_pulse_density(density as i8)
+    }
+
+    /// Set raw prescale
+    ///
+    /// The divider of source clock, ranges from 1 to 256
+    pub fn set_prescale(&self, prescale: u16) -> Result<(), Error> {
+        if (1..=256).contains(&prescale) {
+            GPIO_SD::set_prescale(N, prescale);
+            Ok(())
+        } else {
+            Err(Error::PrescaleRange)
+        }
+    }
+
+    /// Set prescale using frequency
+    pub fn set_frequency(&self, frequency: HertzU32) -> Result<(), Error> {
+        let clocks = Clocks::get();
+        let clock_frequency = clocks.apb_clock.to_Hz();
+        let frequency = frequency.to_Hz();
+
+        let prescale = prescale_from_frequency(clock_frequency, frequency);
+
+        self.set_prescale(prescale)
+    }
+}
+
+mod ehal1 {
+    use embedded_hal::pwm::{Error as PwmError, ErrorKind, ErrorType, SetDutyCycle};
+
+    use super::{Channel, Error, OutputPin};
+
+    impl PwmError for Error {
+        fn kind(&self) -> ErrorKind {
+            ErrorKind::Other
+        }
+    }
+
+    impl<'d, const N: u8, O: OutputPin> ErrorType for Channel<'d, N, O> {
+        type Error = Error;
+    }
+
+    impl<'d, const N: u8, O: OutputPin> SetDutyCycle for Channel<'d, N, O> {
+        fn max_duty_cycle(&self) -> u16 {
+            255
+        }
+
+        fn set_duty_cycle(&mut self, mut duty: u16) -> Result<(), Self::Error> {
+            let max = self.max_duty_cycle();
+            duty = if duty > max { max } else { duty };
+            self.set_duty(duty as u8);
+            Ok(())
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+const CHANNELS: [OutputSignal; 8] = [
+    OutputSignal::GPIO_SD0,
+    OutputSignal::GPIO_SD1,
+    OutputSignal::GPIO_SD2,
+    OutputSignal::GPIO_SD3,
+    OutputSignal::GPIO_SD4,
+    OutputSignal::GPIO_SD5,
+    OutputSignal::GPIO_SD6,
+    OutputSignal::GPIO_SD7,
+];
+
+#[cfg(any(esp32c3, esp32c6, esp32h2))]
+const CHANNELS: [OutputSignal; 4] = [
+    OutputSignal::GPIO_SD0,
+    OutputSignal::GPIO_SD1,
+    OutputSignal::GPIO_SD2,
+    OutputSignal::GPIO_SD3,
+];
+
+#[doc(hidden)]
+pub trait RegisterAccess {
```

Comment:
This trait isn't really being used, the methods should just be inlined.

Reply:
Ok